### PR TITLE
avoid calling sc_*_binary recursively

### PR DESCRIPTION
**Pull request by frankmorgner**
- lock the card early to avoid deselection of the file
- check on integer overflows of indices

fixes https://github.com/OpenSC/OpenSC/issues/1919

Since this affects many cards, I'll not integrate this until 0.21.0 is out.

##### Checklist

- [ ] Documentation is added or updated
- [ ] New files have a LGPL 2.1 license statement
- [ ] PKCS#11 module is tested
- [ ] Windows minidriver is tested
- [ ] macOS tokend is tested

### CODE REVIEW

File: src/libopensc/card.c [hunk 1]
```diff
@@ -637,41 +637,39 @@ int sc_read_binary(sc_card_t *card, unsigned int idx,
...
+		count -= (size_t) r;
+		buf   += (size_t) r;
+		idx   += (size_t) r;
 	}
-	r = card->ops->read_binary(card, idx, buf, count, flags);
-	LOG_FUNC_RETURN(card->ctx, r);
+
+	sc_unlock(card);
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
```
Comment (popovec) [inline src/libopensc/card.c]: LOG_FUNC_RETURN(card->ctx, r);

**Comment by popovec** [comment]
OsEID simulator - T0/T1 protocol, max_recv_size 256/255, init card, rsa import keys/sign/decrypt (512..2048) ECDSA/ECDH (NIST192..521).. all OK

**Comment by frankmorgner** [comment]
So the above problem with EF.ODF was fixed with https://github.com/OpenSC/OpenSC/pull/1924/commits/1f9b3f35383e2d7384741684eb4a25b791c9257f, right?

**Comment by popovec** [comment]
91af2c7513b72d561628098b4d178e8ebe358188 solved the problem with EF.ODF,  I did tests with 167c03fde4e5c763af3bfc9a1ab406c012ef627b  - all tests passed.

**Comment by popovec** [comment]
I run new test with modified simulator, read_binary does not return any data but SW 0x9000 (as discussed in #1919). Seems return code from `sc_read_binary()` is still not correct (128 but it should be 0  or SC_ERROR_???).

```
P:31563; T:0x139838377472448 13:43:42.528 [pkcs15-init] apdu.c:370:sc_single_transmit: CLA:0, INS:B0, P1:0, P2:0, data(0) (nil)
P:31563; T:0x139838377472448 13:43:42.528 [pkcs15-init] reader-pcsc.c:316:pcsc_transmit: reader 'OsEIDsim 00 00'
P:31563; T:0x139838377472448 13:43:42.528 [pkcs15-init] reader-pcsc.c:317:pcsc_transmit: 
Outgoing APDU (5 bytes):
00 B0 00 00 80 .....
P:31563; T:0x139838377472448 13:43:42.528 [pkcs15-init] reader-pcsc.c:235:pcsc_internal_transmit: called
P:31563; T:0x139838377472448 13:43:42.528 [pkcs15-init] reader-pcsc.c:326:pcsc_transmit: 
Incoming APDU (2 bytes):
90 00 ..
P:31563; T:0x139838377472448 13:43:42.528 [pkcs15-init] apdu.c:382:sc_single_transmit: returning with: 0 (Success)
P:31563; T:0x139838377472448 13:43:42.529 [pkcs15-init] apdu.c:535:sc_transmit: returning with: 0 (Success)
P:31563; T:0x139838377472448 13:43:42.529 [pkcs15-init] card.c:523:sc_unlock: called
P:31563; T:0x139838377472448 13:43:42.529 [pkcs15-init] iso7816.c:164:iso7816_read_binary: returning with: 0 (Success)
P:31563; T:0x139838377472448 13:43:42.529 [pkcs15-init] card.c:523:sc_unlock: called
P:31563; T:0x139838377472448 13:43:42.529 [pkcs15-init] reader-pcsc.c:730:pcsc_unlock: called
P:31563; T:0x139838377472448 13:43:42.538 [pkcs15-init] card.c:673:sc_read_binary: returning with: 128
P:31563; T:0x139838377472448 13:43:42.538 [pkcs15-init] pkcs15-lib.c:4367:sc_pkcs15init_parse_info: OpenSC info file corrupted
```

**Comment by frankmorgner** [comment]
Indeed...

I've checked all cards' implementations of read/write/update/erase_binary and it turns out, that none returns SC_SUCCESS if everything is fine (with one exception, but I've changed it to be compliant). Anyway, now SC_SUCCESS or a return value of `0` aborts the processing.

**Comment by Jakuje** [comment]
sigh .. this looks like breaking support for IDPrime, probably because it was filled before the idprime branch was merged and I did not check. Let me fill a new PR updating the IDPrime

**Pull request merged**